nym-vpn-core/crates/nym-vpn-account-controller/src/storage.rs

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+use std::sync::Arc;
+
+use nym_compact_ecash::VerificationKeyAuth;
+use nym_credentials::{
+    AggregatedCoinIndicesSignatures, AggregatedExpirationDateSignatures, EpochVerificationKey,
+    IssuedTicketBook,
+};
+use nym_credentials_interface::TicketType;
+use nym_sdk::mixnet::CredentialStorage;
+use nym_vpn_api_client::types::{Device, VpnApiAccount};
+use nym_vpn_store::{keys::KeyStore, mnemonic::MnemonicStorage, VpnStorage};
+
+use crate::error::Error;
+
+pub(crate) struct AccountStorage<S>
+where
+    S: VpnStorage,
+{
+    storage: Arc<tokio::sync::Mutex<S>>,
+}
+
+impl<S> AccountStorage<S>
+where
+    S: VpnStorage,
+{
+    pub(crate) fn from(storage: Arc<tokio::sync::Mutex<S>>) -> Self {
+        Self { storage }
+    }
+
+    // Load account and keep the error type
+    pub(crate) async fn load_account_from_storage(
+        &self,
+    ) -> Result<VpnApiAccount, <S as MnemonicStorage>::StorageError> {
+        self.storage
+            .lock()
+            .await
+            .load_mnemonic()
+            .await
+            .map(VpnApiAccount::from)
+            .inspect(|account| tracing::debug!("Loading account id: {}", account.id()))
+    }
+
+    // Convenience function to load account and box the error
+    pub(crate) async fn load_account(&self) -> Result<VpnApiAccount, Error> {
+        self.load_account_from_storage()
+            .await
+            .map_err(|err| Error::MnemonicStore {
+                source: Box::new(err),
+            })
+    }
+
+    // Load device keys and keep the error type
+    pub(crate) async fn load_device_keys_from_storage(
+        &self,
+    ) -> Result<Device, <S as KeyStore>::StorageError> {
+        self.storage
+            .lock()
+            .await
+            .load_keys()
+            .await
+            .map(|keys| Device::from(keys.device_keypair()))
+            .inspect(|device| {
+                tracing::debug!("Loading device keys: {}", device.identity_key());
+            })
+    }
+
+    // Convenience function to load device keys and box the error
+    pub(crate) async fn load_device_keys(&self) -> Result<Device, Error> {
+        self.load_device_keys_from_storage()
+            .await
+            .map_err(|err| Error::KeyStore {
+                source: Box::new(err),
+            })
+    }
+}
+
+pub(crate) struct VpnCredentialStorage {
+    pub(crate) storage: nym_credential_storage::persistent_storage::PersistentStorage,
+}
+
+impl VpnCredentialStorage {
+    pub(crate) async fn check_local_remaining_tickets(&self) -> Vec<(TicketType, u32)> {
+        // TODO: remove unwrap
+        let ticketbooks_info = self.storage.get_ticketbooks_info().await.unwrap();
+
+        // For each ticketbook type, iterate over and check if we have enough tickets stored
+        // locally
+        let ticketbook_types = ticketbook_types();
+
+        let mut request_zk_nym = Vec::new();
+        for ticketbook_type in ticketbook_types.iter() {
```

Comment:
wouldn't it make more sense to return a `HashMap` with an entry per `TicketType` in case you have multiple ticketbooks for the same underlying ticketbook type?

Reply:
They are summed per ticketbook type. Problem with HashMap is the requirment on deriving Hash, which I encountered. But let me double check